libuhuru/doc/man/fr/uhuru-scan.1.txt
UHURU-SCAN(1)
=============
:doctype: manpage

NOM
---
uhuru-scan - outil de scan en ligne de commande


SYNOPSIS
--------
*uhuru-scan* ['OPTIONS'] 'FILE|DIR'


DESCRIPTION
-----------
La commande *uhuru-scan*(1) lance une recherche de malwares dans le fichier ou répertoire passé en argument et affiche les résultats sur sa sortie standard.

Cette commande utilise la librairie de scan *libuhuru*(3).


OPTIONS
-------
*-r, --recursive*::
    Récursif.
    Analyse récursive des répertoires. Lance la recherche dans le répertoire et dans tous ses sous-répertoires.

*-t, --threaded*::
    Threadé.
    Utilise un ensemble de threads pour accélérer l'analyse.

*-n, --no-summary*::
    Pas de résumé.
    N'affiche pas le résumé de l'analyse à la fin de l'exécution.

*-h, --help*::
    Aide
    Affiche l'aide et termine l'exécution.


EXIT STATUS
-----------
*0*::
    Succès

*1*::
    Erreur


AUTHOR
------
uhuru-scan a été écrit par François Déchelle.


RESOURCES
---------
Site web: <http://uhuru-solutions.com/>


COPYING
-------
Copyright \(C) 2014 qui?. Free use of this software is
granted under the terms of the GNU General Public License (GPL).


gui/ChangeLog
2015-05-29  François  <francois@dechelle.net>

	* implementation of recent scans in progress; almost works

2015-05-28  François  <francois@dechelle.net>

	* adding recent scan menu

2015-05-27  François  <francois@dechelle.net>

	* internationalization completed

2015-05-12  François  <francois@dechelle.net>

	* dbus debug in progress

2015-04-02  François  <francois@dechelle.net>

	* added scan duration in scan dialog

	* finished i18n; replaced systray menu creation with manual menu, no more XML or glade

2015-03-27  François  <francois@dechelle.net>

	* added i18n using intltool; only language added yet is fr

2015-03-16  François  <francois@dechelle.net>

	* added systray implementation with old gtk_status_icon()

	* added non-scalable PNG icons

2015-03-12  François  <francois@dechelle.net>

	* fixed icon path

2015-03-11  François  <francois@dechelle.net>

	* moved interface to GTK; Qt is phased out

2015-03-10  François  <francois@dechelle.net>

	* added about dialog

2015-03-06  François  <francois@dechelle.net>

	* added systray icon and menu; works almost like Qt version (maybe better...)

	* adding file selection dialog

	* fixed state, action & path display

	* fixed some sorting and resizing problems

2015-03-05  François  <francois@dechelle.net>

	* fixed close button

	* added unhandled file count

	* added current scanning path displayed in progress bar

	* fixed progress bar and to scan file count

	* changed implementation of GTK model to use polled I/O instead of threads; works!

2015-02-04  François  <francois@dechelle.net>

	* added implementation of model counters using glib/gobject

2015-02-03  François  <francois@dechelle.net>

	* starting implementation of GTK gui

2015-01-29  François  <francois@dechelle.net>

	* moving Qt gui to its own directory; preparing move to gtk ;-)

2015-01-13  François  <francois@dechelle.net>

	* trying to fix report table column resizing; added new counters

2015-01-08  François  <francois@dechelle.net>

	* fixed configure.ac and systray.cc w.r.t. CentOS packaging

2014-12-26  François  <francois@dechelle.net>

	* removed -rpath option in src/Makefile.am, which made lintian fail

2014-12-10  François  <francois@dechelle.net>

	* fixed make dist

	* added a 'About' menu entry, with associated dialog

2014-12-05  François  <francois@dechelle.net>

	* added a 'Recent scans' menu entry, and fixed initialization of ScanWidget when displaying a passed scan

2014-12-04  François  <francois@dechelle.net>

	* changed scan window size and position

2014-12-03  François  <francois@dechelle.net>

	* added a close button to the scan window, enabled when scan is completed

 	* tested umwsu.cpp w.r.t. load thread

	* changing umwsu.cpp in order to call umwsu_open at startup in a separate thread, in order not to block interface startup

2014-12-01  François  <francois@dechelle.net>

	* added .desktop file; added make install rules for .desktop file and icons

2014-11-25  François  <francois@dechelle.net>

	* improved model to display analysis file names and select them (copy/paste works)

	* temporarily giving up with icon animation; will see later

2014-11-24  François  <francois@dechelle.net>

	* changed scan model to call umwsu_*() inside a QThread; now tables updates continuously

	* implemented correctly the model derived from QAbstractTableModel; now the table updates ok

2014-11-19  François  <francois@dechelle.net>

	* adding QAbstractTableModel derived model for scan report

2014-11-18  François  <francois@dechelle.net>

	* switched to uhuru-linux-gui git repo

	* fixed some compilation and installation problems with automake/autoconf; now runs ok with icon

2014-11-17  François  <francois@dechelle.net>

	* compile and run with automake/autoconf; will make packaging much easier

	* added automake/autoconf stuff

2014-11-05  François  <francois@dechelle.net>

	* changed scan widget components, changed constructors signatures

	* first working version with systray; scans ok

	* added file selection dialog to systray action

	* added systray to IHM

	* reorganized sources; moved origin UI to legacy/

2014-09-24  François Déchelle  <francois@dechelle.net>

	* ported Uhuru/ to QT4: works ok, except strange compile error in scanondemand.cpp




libuhuru/tools/uhuru-scan.c
#include <libuhuru/scan.h>

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/epoll.h>

struct scan_summary {
  int scanned;
  int malware;
  int suspicious;
  int unhandled;
  int clean;
};

struct scan_options {
  int use_daemon;
  int recursive;
  int threaded;
  int no_summary;
  int print_clean;
  /* more options later */
  char *path;
};

static struct option long_options[] = {
  {"help",         no_argument,       0, 'h'},
  {"local",        no_argument,       0, 'l'},
  {"recursive",    no_argument,       0, 'r'},
  {"threaded",     no_argument,       0, 't'},
  {"no-summary",   no_argument,       0, 'n'},
  {"print-clean",  no_argument,       0, 'c'},
  {0, 0, 0, 0}
};

static void usage(void)
{
  fprintf(stderr, "usage: uhuru-scan [options] FILE|DIR\n");
  fprintf(stderr, "\n");
  fprintf(stderr, "Uhuru antivirus scanner\n");
  fprintf(stderr, "\n");
  fprintf(stderr, "Options:\n");
  fprintf(stderr, "  --help  -h               print help and quit\n");
  fprintf(stderr, "  --local  -l              do not use the scan daemon\n");
  fprintf(stderr, "  --recursive  -r          scan directories recursively\n");
  fprintf(stderr, "  --threaded -t            scan using multiple threads\n");
  fprintf(stderr, "  --no-summary -n          disable summary at end of scanning\n");
  fprintf(stderr, "  --print-clean -c         print also clean files as they are scanned\n");
  fprintf(stderr, "\n");

  exit(1);
}

static void parse_options(int argc, char *argv[], struct scan_options *opts)
{
  int c;

  opts->use_daemon = 1;
  opts->recursive = 0;
  opts->threaded = 1;
  opts->no_summary = 0;
  opts->print_clean = 0;
  opts->path = NULL;

  while (1) {
    int option_index = 0;

    c = getopt_long(argc, argv, "hlrtnc", long_options, &option_index);

    if (c == -1)
      break;

    switch (c) {
    case 'h':
      usage();
      break;
    case 'l':
      opts->use_daemon = 0;
      break;
    case 'r':
      opts->recursive = 1;
      break;
    case 't':
      opts->threaded = 1;
      break;
    case 'n':
      opts->no_summary = 1;
      break;
    case 'c':
      opts->print_clean = 1;
      break;
    default:
      usage();
    }
  }

  if (optind < argc)
    opts->path = argv[optind];

  if (opts->path == NULL)
    usage();
}

static void report_print_callback(struct uhuru_report *report, void *callback_data)
{
  int *print_clean = (int *)callback_data;

  if (!*print_clean && (report->status == UHURU_WHITE_LISTED || report->status == UHURU_CLEAN))
    return;

  uhuru_report_print(report, stdout);
}

static void summary_callback(struct uhuru_report *report, void *callback_data)
{
  struct scan_summary *s = (struct scan_summary *)callback_data;

  s->scanned++;

  switch(report->status) {
  case UHURU_MALWARE:
    s->malware++;
    break;
  case UHURU_SUSPICIOUS:
    s->suspicious++;
    break;
  case UHURU_EINVAL:
  case UHURU_IERROR:
  case UHURU_UNKNOWN_FILE_TYPE:
  case UHURU_UNDECIDED:
    s->unhandled++;
    break;
  case UHURU_WHITE_LISTED:
  case UHURU_CLEAN:
    s->clean++;
    break;
  }
}

static void poll_add_fd(int epoll_fd, int fd)
{
  struct epoll_event ev;

  ev.events = EPOLLIN;
  ev.data.fd = fd;
  if (epoll_ctl(epoll_fd, EPOLL_CTL_ADD, fd, &ev) == -1) {
    perror("epoll_ctl");
    exit(EXIT_FAILURE);
  }
}

static void scan_loop_poll(struct uhuru_scan *scan)
{
#define MAX_EVENTS 10
  struct epoll_event events[MAX_EVENTS];
  int epoll_fd;

  epoll_fd = epoll_create(42);
  if (epoll_fd < 0) {
    perror("epoll_create");
    exit(EXIT_FAILURE);
  }

  poll_add_fd(epoll_fd, STDIN_FILENO);
  poll_add_fd(epoll_fd, uhuru_scan_get_poll_fd(scan));

  while (1) {
    int n_ev, n;

    n_ev = epoll_wait(epoll_fd, events, MAX_EVENTS, -1);
    if (n_ev == -1) {
      perror("epoll_wait");
      exit(EXIT_FAILURE);
    }

    for (n = 0; n < n_ev; n++) {
      if (events[n].data.fd == STDIN_FILENO) {
	;
      } else {
	if (uhuru_scan_run(scan) != UHURU_SCAN_CONTINUE)
	  return;
      }
    }
  }
}

static void scan_loop_no_poll(struct uhuru_scan *scan)
{
  while(uhuru_scan_run(scan) == UHURU_SCAN_CONTINUE)
    ;
}

static void do_scan(struct scan_options *opts, struct scan_summary *summary)
{
  enum uhuru_scan_flags flags = 0;
  struct uhuru *u;
  struct uhuru_scan *scan;

  if (opts->recursive)
    flags |= UHURU_SCAN_RECURSE;

  if (opts->threaded)
    flags |= UHURU_SCAN_THREADED;

  u = uhuru_open(opts->use_daemon);

  uhuru_set_verbose(u, 1);

#if 0
  uhuru_print(u);
#endif

  scan = uhuru_scan_new(u, opts->path, flags);

  uhuru_scan_add_callback(scan, report_print_callback, &opts->print_clean);

  if (!opts->no_summary) {
    summary->scanned = summary->malware = summary->suspicious = summary->unhandled = summary->clean = 0;

    uhuru_scan_add_callback(scan, summary_callback, summary);
  }

  uhuru_scan_start(scan);

  if (opts->use_daemon)
    scan_loop_poll(scan);
  else
    scan_loop_no_poll(scan);

  uhuru_scan_free(scan);

  uhuru_close(u);
}

static void print_summary(struct scan_summary *summary)
{
  printf("\nSCAN SUMMARY:\n");
  printf("scanned files     : %d\n", summary->scanned);
  printf("malware files     : %d\n", summary->malware);
  printf("suspicious files  : %d\n", summary->suspicious);
  printf("unhandled files   : %d\n", summary->unhandled);
  printf("clean files       : %d\n", summary->clean);
}

int main(int argc, char **argv)
{
  struct scan_options *opts = (struct scan_options *)malloc(sizeof(struct scan_options));
  struct scan_summary *summary = NULL;

  parse_options(argc, argv, opts);

  if (!opts->no_summary)
    summary = (struct scan_summary *)malloc(sizeof(struct scan_summary));

  do_scan(opts, summary);

  if (!opts->no_summary)
    print_summary(summary);

  return 0;
}
